examples/solenoid_softedge/analysis_solenoid_softedge.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Chad Mitchell, Axel Huebl
+# License: BSD-3-Clause-LBNL
+#
+
+import glob
+
+import numpy as np
+import pandas as pd
+from scipy.stats import moment
+
+
+def get_moments(beam):
+    """Calculate standard deviations of beam position & momenta
+    and emittance values
+
+    Returns
+    -------
+    sigx, sigy, sigt, emittance_x, emittance_y, emittance_t
+    """
+    sigx = moment(beam["x"], moment=2) ** 0.5  # variance -> std dev.
+    sigpx = moment(beam["px"], moment=2) ** 0.5
+    sigy = moment(beam["y"], moment=2) ** 0.5
+    sigpy = moment(beam["py"], moment=2) ** 0.5
+    sigt = moment(beam["t"], moment=2) ** 0.5
+    sigpt = moment(beam["pt"], moment=2) ** 0.5
+
+    epstrms = beam.cov(ddof=0)
+    emittance_x = (sigx**2 * sigpx**2 - epstrms["x"]["px"] ** 2) ** 0.5
+    emittance_y = (sigy**2 * sigpy**2 - epstrms["y"]["py"] ** 2) ** 0.5
+    emittance_t = (sigt**2 * sigpt**2 - epstrms["t"]["pt"] ** 2) ** 0.5
+
+    return (sigx, sigy, sigt, emittance_x, emittance_y, emittance_t)
+
+
+def read_all_files(file_pattern):
+    """Read in all CSV files from each MPI rank (and potentially OpenMP
+    thread). Concatenate into one Pandas dataframe.
+
+    Returns
+    -------
+    pandas.DataFrame
+    """
+    return pd.concat(
+        (
+            pd.read_csv(filename, delimiter=r"\s+")
+            for filename in glob.glob(file_pattern)
+        ),
+        axis=0,
+        ignore_index=True,
+    ).set_index("id")
+
+
+# initial/final beam on rank zero
+initial = read_all_files("diags/beam_000000.*")
+final = read_all_files("diags/beam_final.*")
+
+# compare number of particles
+num_particles = 10000
+assert num_particles == len(initial)
+assert num_particles == len(final)
+
+print("Initial Beam:")
+sigx, sigy, sigt, emittance_x, emittance_y, emittance_t = get_moments(initial)
+print(f"  sigx={sigx:e} sigy={sigy:e} sigt={sigt:e}")
+print(
+    f"  emittance_x={emittance_x:e} emittance_y={emittance_y:e} emittance_t={emittance_t:e}"
+)
+
+atol = 0.0  # ignored
+rtol = 2.0 * num_particles**-0.5  # from random sampling of a smooth distribution
+print(f"  rtol={rtol} (ignored: atol~={atol})")
+
+assert np.allclose(
+    [sigx, sigy, sigt, emittance_x, emittance_y, emittance_t],
+    [
+        1.559531175539e-3,
+        2.205510139392e-3,
+        1.0e-3,
+        1.0e-6,
+        2.0e-6,
+        1.0e-6,
+    ],
+    rtol=rtol,
+    atol=atol,
+)
+
+
+print("")
+print("Final Beam:")
+sigx, sigy, sigt, emittance_x, emittance_y, emittance_t = get_moments(final)
+print(f"  sigx={sigx:e} sigy={sigy:e} sigt={sigt:e}")
+print(
+    f"  emittance_x={emittance_x:e} emittance_y={emittance_y:e} emittance_t={emittance_t:e}"
+)
+
+atol = 0.0  # ignored
+rtol = 2.0 * num_particles**-0.5  # from random sampling of a smooth distribution
```

Comment:
We need to relax this more as seen on Windows: mostly for the emittance values.

Reply:
We see that the Windows test (and only the Windows test) now fails with NaN for some values, so this is not a precision issue, but some OS/compiler-dependent bug.

Reply:
The failure is limited to the Windows tests, and this appears to be a compiler-dependent bug (not a statistical fluctuation due to the initial random sampling).  The output on the two Windows test is very different (eg, a factor of ~2 in sig_y).  On both tests, the initial and final emittance values coincide _exactly._  This does not make sense, because the beam is rotated  by the solenoid, resulting in a coupling of the horizontal and vertical planes.  This suggests that there may be a problem with post-processing of output for the Windows tests.

Reply:
THank you for the analysis. I also see that the tests on Windows only fail for the inputs-file driven case, not for the Python-equivalent one.

Reply:
pushed a fix here as well